plugins/producer_plugin/test/test_block_timing_util.cpp

```diff
@@ -34,73 +35,214 @@ BOOST_AUTO_TEST_CASE(test_calculate_block_deadline) {
+      BOOST_CHECK_EQUAL(calculate_producing_block_deadline(cpu_effort_us, third_block_time),
                         third_block_time.to_time_point() - fc::milliseconds(300));
 
       // use previous deadline as now
       fc::mock_time_traits::set_now(third_block_time.to_time_point() - fc::milliseconds(300));
       auto forth_block_time = eosio::chain::block_timestamp_type(production_round_1st_block_slot + 3);
-      BOOST_CHECK_EQUAL(calculate_block_deadline(cpu_effort_us, eosio::pending_block_mode::producing, forth_block_time),
+      BOOST_CHECK_EQUAL(calculate_producing_block_deadline(cpu_effort_us, forth_block_time),
                         forth_block_time.to_time_point() - fc::milliseconds(400));
 
       ///////////////////////////////////////////////////////////////////////////////////////////////////
 
       auto seventh_block_time = eosio::chain::block_timestamp_type(production_round_1st_block_slot + 6);
       fc::mock_time_traits::set_now(seventh_block_time.to_time_point() - fc::milliseconds(500));
 
-      BOOST_CHECK_EQUAL(calculate_block_deadline(cpu_effort_us, eosio::pending_block_mode::producing, seventh_block_time),
+      BOOST_CHECK_EQUAL(calculate_producing_block_deadline(cpu_effort_us, seventh_block_time),
                         seventh_block_time.to_time_point() - fc::milliseconds(100));
 
       // use previous deadline as now
       fc::mock_time_traits::set_now(seventh_block_time.to_time_point() - fc::milliseconds(100));
       auto eighth_block_time = eosio::chain::block_timestamp_type(production_round_1st_block_slot + 7);
 
-      BOOST_CHECK_EQUAL(calculate_block_deadline(cpu_effort_us, eosio::pending_block_mode::producing, eighth_block_time),
+      BOOST_CHECK_EQUAL(calculate_producing_block_deadline(cpu_effort_us, eighth_block_time),
                         eighth_block_time.to_time_point() - fc::milliseconds(200));
 
       // use previous deadline as now
       fc::mock_time_traits::set_now(eighth_block_time.to_time_point() - fc::milliseconds(200));
       auto ninth_block_time = eosio::chain::block_timestamp_type(production_round_1st_block_slot + 8);
 
-      BOOST_CHECK_EQUAL(calculate_block_deadline(cpu_effort_us, eosio::pending_block_mode::producing, ninth_block_time),
+      BOOST_CHECK_EQUAL(calculate_producing_block_deadline(cpu_effort_us, ninth_block_time),
                         ninth_block_time.to_time_point() - fc::milliseconds(300));
    }
 }
 
+BOOST_AUTO_TEST_CASE(test_calculate_producer_wake_up_time) {
+   using namespace eosio;
+   using namespace eosio::chain;
+   using namespace eosio::chain::literals;
+   using namespace eosio::block_timing_util;
+
+   producer_watermarks empty_watermarks;
+   // use full cpu effort for most of these tests since calculate_producing_block_deadline is tested above
+   constexpr uint32_t full_cpu_effort = eosio::chain::config::block_interval_us;
+
+   // no producers
```

Comment:
You can add a pair of `{` and `}` around this first test to be consistent with other tests.